compiler/src/yul/mappers/contracts.rs

```diff
@@ -39,28 +31,34 @@ pub fn contract_def(
                     init = Some((
                         functions::func_def(context, stmt)?,
                         attributes.param_types.clone(),
-                        runtime.clone(),
                     ))
                 } else {
                     user_functions.push(functions::func_def(context, stmt)?)
                 }
             }
         }
 
-        let mut statements = vec![];
-        statements.append(&mut user_functions);
-        statements.append(&mut runtime);
-        statements.push(runtime_abi::dispatcher(
-            attributes.public_functions.to_owned(),
-        )?);
+        let runtime = {
+            let mut runtime = runtime_functions::std();
+            runtime.append(&mut build_runtime_functions(
+                attributes.runtime_operations.to_owned(),
+            ));
+            runtime.append(&mut user_functions);
+            runtime.push(runtime_abi::dispatcher(
+                attributes.public_functions.to_owned(),
+            )?);
+            runtime
+        };
 
         return Ok(yul::Object {
             name: identifier! { Contract },
-            code: constructor::build(init),
+            code: constructor::build(init, runtime.clone()),
             objects: vec![yul::Object {
                 name: identifier! { runtime },
                 code: yul::Code {
-                    block: yul::Block { statements },
+                    block: yul::Block {
+                        statements: runtime,
+                    },
```

Comment:
I remember we talked before about this and you said that giving access to the user defined functions in the constructor would mean that we would have to duplicate each of these functions. Is this what we do now? If yes, then we should probably add a github issue to track a possible optimization (but maybe the YUL optimizer will optimize it away for us anyway...who knows)

Reply:
> Is this what we do now?

yeah

> but maybe the YUL optimizer will optimize it away

Compiling the following without the optimizer

```javascript
object "Foo" {
  code {
    function bar() {
      mstore(0, 1)
    }

    mstore(2,3)
  }
}
```

results in

```
======= test.yul (EVM) =======

Pretty printed source:
object "Foo" {
    code {
        function bar()
        { mstore(0, 1) }
        mstore(2, 3)
    }
}


Binary representation:
600b565b60016000525b565b6003600252

Text representation:
    /* "test.yul":28:69   */
  jump(tag_1)
tag_2:
    /* "test.yul":61:62   */
  0x01
    /* "test.yul":58:59   */
  0x00
    /* "test.yul":51:63   */
  mstore
    /* "test.yul":43:69   */
tag_3:
  jump
tag_1:
    /* "test.yul":84:85   */
  0x03
    /* "test.yul":82:83   */
  0x02
    /* "test.yul":75:86   */
  mstore
```

whereas compiling with the optimizer results in

```
======= test.yul (EVM) =======

Pretty printed source:
object "Foo" {
    code { { mstore(2, 3) } }
}


Binary representation:
6003600252

Text representation:
    /* "test.yul":84:85   */
  0x03
    /* "test.yul":82:83   */
  0x02
    /* "test.yul":75:86   */
  mstore
```

so it looks like all of our unused functions will get removed.